### PR TITLE
Fix for duplicate layers due to adding graphic from cc libraries.

### PR DESCRIPTION
**Pull request by shaoshing**
This PR fixes duplicate layers in our LayerStructure model due to adding a graphic with ALT key pressed. Usually, when the ALT key is pressed, PS will add the layers contained in the graphic, instead of creating LLSO, and will not enter modal tool state. However, if the graphic contains only one layer and it is an embedded SO, PS will enter modal state tool, which causes adding the duplicate layer when we handle the `placeEvent`.

So the solution is to check whether the layer is existed when handling `placeEvent`. I also added cheker to throw error in the `layers.addLayer`, if we are trying to add duplicate layers. 

To test: drag graphic `Embedded` from library `Test - OK ...` (or import the graphic from http://adobe.ly/1Mh2XAM) onto the canvas while holding the ALT key.

Addresses issues: #3145 #3146 

While testing, you may see the following error, which is because the layer is not ready for reading (same problem again, and I remember we have watson for it) but it is totally ok for this case because we will add the layer later when handling the `placeEvent`.

screen shot 2015-10-27 at 3 21 50 pm

### CODE REVIEW

**Comment by chadrolfs** [comment]
I want to include this in the PR 3047 push as well

File: src/js/actions/libraries.js [hunk 1]
```diff
@@ -850,6 +850,7 @@ define(function (require, exports) {
                         //
                         // FIXME: we should instead get IDs back from Photoshop when layers are placed with modifier, 
                         //        so we don't have to get all the layer IDs.
+                        //        https://watsonexp.corp.adobe.com/#bug=4080071
```
Comment (shaoshing) [inline src/js/actions/libraries.js]: a new watson bug that could free us from this hack.

**Comment by pineapplespatula** [comment]
This looks good to me @chadrolfs have you tested this branch already?

**Comment by iwehrman** [comment]
Can someone please repro this bug in compiled mode to check whether or not we can recover from it automatically? This is in case we don't get BRC approval for the PR. Or, if we _don't_ recover automatically, that will help make a stronger case to the BRC.

**Comment by shaoshing** [comment]
I haven't test this in compiled mode. But adding duplicate layers is not going to throw any errors, and you will see the swap button (and other buttons involve 2 layers) enabled if you selected the layer, and also a bunch of warnings when the layers panel re-rendered. However, if you click to swap, DS will throw error, and that gives us a chance to recover everything. Base on this, I would say this is *not recoverable*

**Comment by iwehrman** [comment]
It sounds like you're saying that, once the swap action fails it _will_ recover. Is there some scenario in which we explicitly do not recover? Don't get me wrong, I definitely want to fix this :grimacing: I just want to confirm that this is in the class, "temporarily out of date layers panel which will fix itself once the inconsistency causes a problem."

File: src/js/actions/layers.js [hunk 2]
```diff
@@ -381,6 +381,14 @@ define(function (require, exports) {
         if (typeof layerSpec === "number") {
             layerSpec = [layerSpec];
         }
+        
+        layerSpec.forEach(function (layerID) {
+            var layer = document.layers.byID(layerID);
+
+            if (layer) {
+                throw new Error("Trying to add a layer that is already existed: " + layerID + " " + layer.name);
```
Comment (iwehrman) [inline src/js/actions/layers.js]: Grammar nit: this should either read "Trying to add a layer that already exists" or "Trying to add a layer this is already present".

**Comment by iwehrman** [comment]
The patch looks great, besides the grammar nit. If we can get this past the BRC, I am definitely in favor of merging it.

**Comment by shaoshing** [comment]
Yes, then it falls into the class "temporarily out of date layers panel which will fix itself once the inconsistency causes a problem."

**Comment by iwehrman** [comment]
I think we've decided against taking this for M3, but we surely want it in master!

**Pull request merged**